File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.swift
class Solution {
    
    func getReplacebleCharCount(_  dict:  [Character: Int], _ start: Int, _ end: Int)  -> Int{
        var mx = 0 
        for (key, val) in dict{
            mx = max(mx, val)
        }
        
        return (end - start + 1 ) - mx 
    }
    
    func characterReplacement(_ s: String, _ k: Int) -> Int {
        var windowStart = 0, mx = 0
        var inputString = Array(s)
        var dict = [Character: Int]()
        
        for windowEnd in 0..<inputString.count{
            dict[inputString[windowEnd], default: 0 ] += 1
            let replaceCharCount = getReplacebleCharCount(dict, windowStart, windowEnd)
            if replaceCharCount <= k{
                mx = max(mx, windowEnd - windowStart + 1 )
            }else{
                while(getReplacebleCharCount(dict, windowStart, windowEnd) > k){
                    dict[inputString[windowStart], default: 0 ] -= 1
                    windowStart += 1
                }
            }
    
        }
        return mx
    }
}

File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.cpp
class Solution {
public:
    bool isValid(string s, unordered_map<char, int >& mp, int range, int k){
        
        int mx = 0;
        
        for(char ch = 'A'; ch <= 'Z'; ch ++ ){
            if(mp[ch] >=  mx){
                mx = mp[ch];
            }
        }
        
        if( (range - mx)  <= k) return true;
        else return false;
        
    }
    
    
    int characterReplacement(string s, int k) {
        unordered_map<char, int > mp;
        
        int ans = 0, windowStart = 0, sz = s.size() ;
        
        
        for(int windowEnd = 0; windowEnd < sz; windowEnd ++ ){
             char curChar = s[windowEnd];
             mp[curChar]++;
            
            int range = windowEnd - windowStart + 1;
            
            while(windowStart < windowEnd && !isValid(s, mp, range, k)){
                char prevChar = s[windowStart];
                mp[prevChar] -- ;
                windowStart ++ ;
                range = windowEnd - windowStart + 1;
            }
            
            ans = max(ans, windowEnd - windowStart + 1);
        }
        
        return ans;
        
        
    }
};


// C

// 1BBBA